src/qibo/models/error_mitigation.py

```diff
@@ -320,13 +372,144 @@ def vnCDR(
+        if "calibration_matrix" in readout.keys():
+            circuit_result = apply_readout_mitigation(
+                circuit_result, readout["calibration_matrix"]
+            )
+        expval = circuit_result.expectation_from_samples(observable)
+        if "ncircuits" in readout.keys():
+            expval /= circuit_result_cal.expectation_from_samples(observable)
+        val.append(expval)
     mit_val = model(np.array(val).reshape(-1, 1), *optimal_params[0])[0]
     # Return data
     if full_output == True:
         return mit_val, val, optimal_params, train_val
     else:
         return mit_val
+
+
+def get_calibration_matrix(nqubits, backend=None, noise_model=None, nshots=1000):
+    """Computes the calibration matrix for readout mitigation.
+
+    Args:
+        nqubits (int): Total number of qubits.
+        backend (qibo.backends.abstract.Backend): Calculation engine.
+        noise_model (int): (qibo.noise.NoiseModel): Noise model used for simulating noisy computation. This matrix can be used to mitigate the effect of `qibo.noise.ReadoutError`.
+        nshots (int): Number of shots.
+
+    Returns:
+        np.ndarray : The computed (`nqubits`, `nqubits`) calibration matrix for readout mitigation.
+    """
+
+    if backend is None:  # pragma: no cover
+        from qibo.backends import GlobalBackend
+
+        backend = GlobalBackend()
+
+    matrix = np.zeros((2**nqubits, 2**nqubits))
+
+    string = "{0:0" + str(nqubits) + "b}"
+    for i in range(2**nqubits):
+        state = string.format(i)
+        circuit = Circuit(nqubits, density_matrix=True)
+        for q, bit in enumerate(state):
+            if bit == "1":
+                circuit.add(gates.X(q))
+        circuit.add(gates.M(*range(nqubits)))
+        if noise_model != None and backend.name != "qibolab":
+            circuit = noise_model.apply(circuit)
+        freq = backend.execute_circuit(circuit, nshots=nshots).frequencies()
+        column = np.zeros(2**nqubits)
+        for key in freq.keys():
+            f = freq[key] / nshots
+            column[int(key, 2)] = f
+        matrix[:, i] = column
+    return np.linalg.inv(matrix)
+
+
+def apply_readout_mitigation(state, calibration_matrix):
+    """Updates the frequencies of the input state with the mitigated ones obtained with `calibration_matrix`*`state.frequencies()`.
+
+    Args:
+        state (qibo.states.CircuitResult): Input state to be updated.
+        calibration_matrix (np.ndarray): Calibration matrix for readout mitigation.
+
+    Returns:
+        qibo.states.CircuitResult : The input state with the updated frequencies.
+    """
+    freq = np.zeros(2**state.nqubits)
+    for k, v in state.frequencies().items():
+        freq[int(k, 2)] = v
+    freq = freq.reshape(-1, 1)
+    for i, val in enumerate(calibration_matrix @ freq):
+        state._frequencies[i] = float(val)
+    return state
+
+
+def apply_temme_readout_mitigation(
```

Comment:
I would rename this as something like `apply_randomized_readout_mitigation`, for a lack of a better name